### ISSUE TITLE
1-line SEGV 5.004-5.7.0: DESTROY + local

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3360]
Migrated from [rt.perl.org#5712](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=5712) (status was 'resolved')

Searchable as RT5712$

**Comment by p5pRT** [comment]
### From jtobey@john-edwin-tobey.org



This goes into a stack\-eating\, segfaulting loop on every version of
Perl that I have \(5\.004 to perl@&#8203;8269\, including 5\.6\.1\-TRIAL2\)&#8203;:

&nbsp;    perl \-e 'DESTROY \{ local $o \} local $o = bless \[\]; die;'

Here's a version that segfaults without eating stack&#8203;:

&nbsp;    perl \-e 'DESTROY \{ local $o \} eval \{ local $o = bless \[\]; die \}'

Best regards\.
\-John

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.6.1:

Configured by jtobey at Mon Feb  5 04:34:50 EST 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.2.17, archname=i686-linux
    uname='linux hofstadter 2.2.17 #3 fri oct 27 22:57:04 edt 2000 i686 unknown '
    config_args='-Doptimize=-g -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.2 20000220 (Debian GNU/Linux)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    v5.6.1-TRIAL2


@INC for perl v5.6.1:
    /usr/local/lib/perl5/5.6.1/i686-linux
    /usr/local/lib/perl5/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.1/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/jtobey
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/jtobey/bin:/jtobey/local/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/local/sbin:/usr/sbin:/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

The patch below fixes bugs 20001115\.002 and 20010205\.006\.  Patched
against 5\.6\.1\-TRIAL2\.

\-John

<details><summary>Inline Patch</summary>

```diff
--- sv.c~	Wed Jan 31 10:56:50 2001
+++ sv.c	Mon Feb  5 16:41:49 2001
@@ -3712,14 +3712,18 @@
 		if (destructor) {
 		    ENTER;
 		    PUSHSTACKi(PERLSI_DESTROY);
-		    SvRV(&tmpref) = SvREFCNT_inc(sv);
+		    /* Give sv a refcount of at least 2 - fixes 20010205.006 -
+		       perl -e 'DESTROY { local $o } local $o = bless []; exit'
+		    */
+		    SvREFCNT(sv) += 2;
+		    SvRV(&tmpref) = sv;
 		    EXTEND(SP, 2);
 		    PUSHMARK(SP);
 		    PUSHs(&tmpref);
 		    PUTBACK;
 		    call_sv((SV*)GvCV(destructor),
 			    G_DISCARD|G_EVAL|G_KEEPERR);
-		    SvREFCNT(sv)--;
+		    SvREFCNT(sv) -= 2;
 		    POPSTACK;
 		    SPAGAIN;
 		    LEAVE;
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

I like the fix below better\.  It catches some cases that the other one
missed\, such as\.\.\.

&nbsp;    perl \-e 'DESTROY \{\{local $x\} local $x\} local $x=bless \[\]; exit'

However\, you can always segfault with

&nbsp;    perl \-e 'DESTROY \{local $x\[0\]\} local @&#8203;x=\(bless \[\]\); exit'

The basic problem is that perl destroys pieces of data \(makes them
invalid by dropping their refcount to 0\) and does not remove them from
visibility until after it runs user code \(DESTROY\)\.  DESTROY manages
to inc the refcount by putting an item on the save stack with local\.
Then the refcount drops again after DESTROY\, and the thing is freed
twice\.

In the case of 'DESTROY \{ local $x \} local $x = bless \[\]; exit' I
think either save\_scalar\(\) should notice that the thing it's saving
has refcnt 0 and replace it with NULL\, or even better\, the restore\_sv
code in leave\_scope\(\) should have noticed when the original $x's
refcnt dropped to 0 and nulled out the GvSV entry\.

I tried redefining SvREFCNT\_dec\(sv\) from

&nbsp;    \#define SvREFCNT\_dec\(sv\)    sv\_free\(\(SV\*\)sv\)

to
&nbsp;    \#define SvREFCNT\_dec\(sv\)    sv\_free\_ptr\(\(SV\*\*\)&\(sv\)\)

where sv\_free\_ptr\(\) made sv = 0 if sv was freed\.  It compiled with
only a few adjustments\, but I think one of the tests failed\, and it
didn't solve my problem\, and anyway\, it's probably too big a
performance hit\.

Other approaches occurred to me too\, but it is too late at night to
enumerate them\.

I'm a little concerned about the performance impact of this change
too\.  Let me know if you want it\, in which case I'll devise testsuite
additions\.

\-John

<details><summary>Inline Patch</summary>

```diff
--- sv.c.dist	Wed Jan 31 10:56:50 2001
+++ sv.c	Tue Feb  6 03:03:59 2001
@@ -3798,8 +3798,13 @@
 	if (SvROK(sv)) {
 	    if (SvWEAKREF(sv))
 	        sv_del_backref(sv);
-	    else
-	        SvREFCNT_dec(SvRV(sv));
+	    else if (PL_in_clean_all)
+		SvREFCNT_dec(SvRV(sv));
+	    else {
+		SvREFCNT(sv)++;
+		SvREFCNT_dec(SvRV(sv));
+		SvREFCNT(sv)--;
+	    }
 	}
 	else if (SvPVX(sv) && SvLEN(sv))
 	    Safefree(SvPVX(sv));
```
</details>
